src/LinuxTracing/LeafFunctionCallManager.cpp

```diff
@@ -0,0 +1,103 @@
+// frames are executable).
+// Note that we cannot simply set libunwindstack to unwind always two frames and compare the outer
+// frame with the respective one in the callchain carried by the perf_event_open event, as in case
+// of uprobes overriding the return addresses, both addresses would be identical even if the actual
+// addresses (after uprobe patching) are not.
+// More (internal) documentation on this in: go/stadia-orbit-leaf-frame-pointer
+Callstack::CallstackType LeafFunctionCallManager::PatchLeafFunctionCaller(
+    CallchainSamplePerfEvent* event, LibunwindstackMaps* current_maps,
+    LibunwindstackUnwinder* unwinder) {
+  CHECK(event != nullptr);
+  CHECK(current_maps != nullptr);
+  CHECK(unwinder != nullptr);
+  CHECK(event->GetCallchainSize() > 2);
+
+  uint64_t stack_size =
+      event->GetRegisters()[PERF_REG_X86_BP] - event->GetRegisters()[PERF_REG_X86_SP];
+  if (stack_size > SAMPLE_STACK_USER_SIZE_512BYTES) {
+    return Callstack::kFramePointerDwarfStackTooSmallError;
+  }
+
+  const LibunwindstackResult& libunwindstack_result =
+      unwinder->Unwind(event->GetPid(), current_maps->Get(), event->GetRegisters(),
+                       event->GetStackData(), stack_size, true);
+  const std::vector<unwindstack::FrameData>& libunwindstack_callstack =
+      libunwindstack_result.frames();
+
+  if (libunwindstack_callstack.empty()) {
+    ERROR("Discarding sample as DWARF-based unwinding resulted in empty callstack (error: %s)",
+          LibunwindstackUnwinder::LibunwindstackErrorString(libunwindstack_result.error_code()));
+    return Callstack::kDwarfUnwindingError;
+  }
+
+  // In case of an intact frame pointer, the region from $rsp to $rbp will only include the locals
+  // of the current frame, and NOT the return address. Thus, unwinding will only report one frame
+  // (the instruction pointer) and we know that the original callchain is already correct.
+  if (libunwindstack_callstack.size() == 1) {
+    return Callstack::kComplete;
+  }
+
+  // If unwinding results in more than two frames, $rbp was also not set correctly by the caller,
+  // thus frame pointers are not all in non-leaf functions, and our assumptions do not hold.
+  if (libunwindstack_callstack.size() > 2) {
+    return Callstack::kFramePointerUnwindingError;
+  }
+
+  const std::vector<uint64_t> original_callchain = event->ips;
+  CHECK(original_callchain.size() > 2);
+
+  std::vector<uint64_t> result;
+  result.reserve(original_callchain.size() + 1);
+  for (size_t i = 0; i < 2; ++i) {
+    result.push_back(original_callchain[i]);
+  }
+
+  CHECK(libunwindstack_callstack.size() == 2);
+  uint64_t libunwindstack_leaf_caller_pc = libunwindstack_callstack[1].pc;
+
+  // If the caller is not executable, we have an unwinding error.
+  unwindstack::MapInfo* map_info = current_maps->Find(libunwindstack_leaf_caller_pc);
+  if (map_info == nullptr || (map_info->flags & PROT_EXEC) == 0) {
+    return Callstack::kDwarfUnwindingError;
```

Comment:
Same thought here about possibly having a separate type for DWARF errors happening in this function.

Reply:
Added a `kStackTopDwarfUnwindingError`